File: patterns/01-design-patterns/03-behavioral/04-command/go/command.go
package command

type PatternError struct{ code string }

func (e *PatternError) Error() string { return e.code }
func (e *PatternError) Code() string  { return e.code }

type RouteTable struct{ routes map[string]string }
type Command interface {
	Execute()
	Undo()
}
type setRoute struct {
	receiver             *RouteTable
	key, value, previous string
	existed              bool
}

func (c *setRoute) Execute() {
	c.previous, c.existed = c.receiver.routes[c.key]
	c.receiver.routes[c.key] = c.value
}
func (c *setRoute) Undo() {
	if c.existed {
		c.receiver.routes[c.key] = c.previous
	} else {
		delete(c.receiver.routes, c.key)
	}
}

type Invoker struct{ history []Command }

func (i *Invoker) Run(c Command) { c.Execute(); i.history = append(i.history, c) }
func (i *Invoker) Undo() error {
	if len(i.history) == 0 {
		return &PatternError{code: "no_history"}
	}
	last := i.history[len(i.history)-1]
	i.history = i.history[:len(i.history)-1]
	last.Undo()
	return nil
}
func Evaluate(input map[string]any) (any, error) {
	routes := map[string]string{}
	if initial, ok := input["initial"].(map[string]any); ok {
		for k, v := range initial {
			routes[k] = v.(string)
		}
	}
	table := &RouteTable{routes}
	invoker := &Invoker{}
	raw, _ := input["commands"].([]any)
	for _, item := range raw {
		v := item.(map[string]any)
		invoker.Run(&setRoute{receiver: table, key: v["key"].(string), value: v["value"].(string)})
	}
	undo := int(input["undo"].(float64))
	for n := 0; n < undo; n++ {
		if err := invoker.Undo(); err != nil {
			return nil, err
		}
	}
	out := map[string]any{}
	for k, v := range table.routes {
		out[k] = v
	}
	return map[string]any{"routes": out, "history": len(invoker.history)}, nil
}

File: patterns/01-design-patterns/03-behavioral/04-command/python/command.py
class PatternError(Exception):
    def __init__(self,code):super().__init__(code);self.code=code
class RouteTable:
    def __init__(self,initial):self.routes=dict(initial)
class SetRoute:
    def __init__(self,receiver,key,value):self.receiver=receiver;self.key=key;self.value=value;self.existed=False;self.previous=None
    def execute(self):self.existed=self.key in self.receiver.routes;self.previous=self.receiver.routes.get(self.key);self.receiver.routes[self.key]=self.value
    def undo(self):
        if self.existed:self.receiver.routes[self.key]=self.previous
        else:self.receiver.routes.pop(self.key,None)
class Invoker:
    def __init__(self):self.history=[]
    def run(self,command):command.execute();self.history.append(command)
    def undo(self):
        if not self.history:raise PatternError("no_history")
        self.history.pop().undo()
def evaluate(input_data):
    table=RouteTable(input_data.get("initial",{}));invoker=Invoker()
    for value in input_data.get("commands",[]):invoker.run(SetRoute(table,value["key"],value["value"]))
    for _ in range(input_data.get("undo",0)):invoker.undo()
    return{"routes":dict(sorted(table.routes.items())),"history":len(invoker.history)}
